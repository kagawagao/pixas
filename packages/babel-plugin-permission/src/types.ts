import { NodePath, PluginPass } from '@babel/core';
import { SourceLocation, StringLiteral } from '@babel/types';

export interface ExtractSymbols {
  single?: string[];
  multiple?: string[];
  components?: string[];
}

export type PermissionDescriptor = Record<string, any>;

export interface State extends PluginPass {
  permissions: ExtractedPermissionDescriptor[];
  componentNames: string[];
  functionNames: string[];
}

export type ExtractedPermissionDescriptor = PermissionDescriptor & Partial<SourceLocation> & { file?: string };

export type PermissionDescriptorPath = Record<keyof PermissionDescriptor, NodePath<StringLiteral> | undefined>;

export interface Options {
  symbols?: ExtractSymbols;
  properties?: string[];
  removeProperties?: string[];
  onPermissionExtracted?: (filePath: string, permissions: PermissionDescriptor[]) => void;
}

export interface PermissionPluginPass extends PluginPass {
  permissions: ExtractedPermissionDescriptor[];
  componentNames: string[];
  singleFunctionNames: string[];
  multipleFunctionNames: string[];
  props: string[];
  removeProps: string[];
}
