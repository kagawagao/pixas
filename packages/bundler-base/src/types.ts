export interface StartOptions {
  port?: number;
  host?: string;
  open?: boolean;
}

export interface IBundler {
  start: (options?: StartOptions) => Promise<void> | void;
  build: () => Promise<void> | void;
}
