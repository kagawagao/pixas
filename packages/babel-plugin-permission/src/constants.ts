import { ExtractSymbols } from './types';

/**
 * Default symbols to look for in the code.
 */
export const defaultSymbols: ExtractSymbols = {
  single: ['definePermission'],
  multiple: ['definePermissions'],
  components: ['UserPermissionAuthorize'],
};

/**
 * Default properties to extract from the permission descriptor.
 */
export const defaultProps = ['code', 'name', 'value', 'description', 'parent', 'type', 'groups', 'data', 'uris'];

/**
 * Properties that should be removed from the permission descriptor.
 */
export const defaultRemoveProps = ['name', 'value', 'description', 'parent', 'type', 'groups', 'data', 'uris'];

/**
 * Symbol used to mark a permission as extracted.
 */
export const extractedSymbol = Symbol('UserPermissionExtracted');
