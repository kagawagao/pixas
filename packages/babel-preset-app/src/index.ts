import { declare } from '@babel/helper-plugin-utils';

module.exports = declare((api, options) => {
  api.assertVersion(7);

  const isNode = api.caller((caller) => {
    if (caller?.name === 'babel-loader') {
      return (caller as any).target === 'node';
    }
    return false;
  });

  return {
    presets: [
      [
        require.resolve('@babel/preset-env'),
        {
          modules: false, // close this to make tree-shaking work
          useBuiltIns: 'usage',
          corejs: 3,
          ...options,
        },
      ],
      require.resolve('@babel/preset-typescript'),
      [
        require.resolve('@babel/preset-react'),
        {
          runtime: 'automatic',
        },
      ],
    ],
    env: {
      development: {
        plugins: [
          [
            require.resolve('babel-plugin-import'),
            {
              libraryName: 'antd',
              libraryDirectory: 'es',
              style: false,
            },
            'antd',
          ],
          [
            require.resolve('babel-plugin-import'),
            {
              libraryName: 'lodash',
              libraryDirectory: '',
              camel2DashComponentName: false,
            },
            'lodash',
          ],
          [
            require.resolve('babel-plugin-import'),
            {
              libraryName: 'lodash-es',
              libraryDirectory: '',
              camel2DashComponentName: false,
            },
            'lodash-es',
          ],
          [
            require.resolve('@babel/plugin-transform-runtime'),
            {
              useESModules: true,
            },
          ],
          !isNode && require.resolve('react-refresh/babel'),
        ].filter(Boolean),
      },
      test: {
        presets: [
          [
            require.resolve('@babel/preset-env'),
            {
              useBuiltIns: 'usage',
              corejs: 3,
            },
          ],
          require.resolve('@babel/preset-typescript'),
          [
            require.resolve('@babel/preset-react'),
            {
              runtime: 'automatic',
            },
          ],
        ],
        plugins: [
          [require.resolve('babel-plugin-import'), false, 'antd'],
          [require.resolve('babel-plugin-import'), false, 'wau'],
          require.resolve('@babel/plugin-transform-runtime'),
        ],
      },
      production: {
        plugins: [
          [
            require.resolve('babel-plugin-import'),
            {
              libraryName: 'antd',
              libraryDirectory: 'es',
              style: false,
            },
            'antd',
          ],
          [
            require.resolve('babel-plugin-import'),
            {
              libraryName: 'antd-mobile',
              libraryDirectory: 'es',
              style: false,
            },
            'antd-mobile',
          ],
          [
            require.resolve('babel-plugin-import'),
            {
              libraryName: 'lodash',
              libraryDirectory: '',
              camel2DashComponentName: false,
            },
            'lodash',
          ],
          [
            require.resolve('babel-plugin-import'),
            {
              libraryName: 'lodash-es',
              libraryDirectory: '',
              camel2DashComponentName: false,
            },
            'lodash-es',
          ],
          [
            require.resolve('@babel/plugin-transform-runtime'),
            {
              useESModules: true,
            },
          ],
        ],
      },
    },
  } as any;
});
