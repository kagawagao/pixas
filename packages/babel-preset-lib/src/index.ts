import { declare } from '@babel/helper-plugin-utils';

const targets = {
  chrome: 62,
  // ie: 11,
  node: 16,
};

module.exports = declare((api, options) => {
  api.assertVersion(7);
  return {
    presets: [
      [
        require.resolve('@babel/preset-env'),
        {
          useBuiltIns: 'usage',
          modules: false,
          corejs: 3,
          targets,
          ...options,
        },
      ],
      require.resolve('@babel/preset-typescript'),
      [
        require.resolve('@babel/preset-react'),
        {
          runtime: 'automatic',
        },
      ],
    ],
    plugins: [
      [
        require.resolve('@babel/plugin-transform-runtime'),
        {
          useESModules: true,
        },
      ],
      [
        'import',
        {
          libraryName: 'antd',
          libraryDirectory: 'es',
          style: true,
        },
        'antd',
      ],
    ],
    env: {
      commonjs: {
        presets: [
          [
            require.resolve('@babel/preset-env'),
            {
              useBuiltIns: 'usage',
              corejs: 3,
              targets,
              ...options,
            },
          ],
          require.resolve('@babel/preset-typescript'),
          [
            require.resolve('@babel/preset-react'),
            {
              runtime: 'automatic',
            },
          ],
        ],
        plugins: [
          [require.resolve('@babel/plugin-transform-runtime')],
          [
            'import',
            {
              libraryName: 'antd',
              libraryDirectory: 'lib',
              style: true,
            },
            'antd',
          ],
        ],
      },
      test: {
        presets: [
          [
            require.resolve('@babel/preset-env'),
            {
              useBuiltIns: 'usage',
              corejs: 3,
              targets,
              ...options,
            },
          ],
          require.resolve('@babel/preset-typescript'),
          [
            require.resolve('@babel/preset-react'),
            {
              runtime: 'automatic',
            },
          ],
        ],
        plugins: [
          [require.resolve('@babel/plugin-transform-runtime')],
          [
            'import',
            {
              libraryName: 'antd',
              libraryDirectory: 'lib',
              style: false,
            },
            'antd',
          ],
        ],
      },
    },
  } as any;
});
