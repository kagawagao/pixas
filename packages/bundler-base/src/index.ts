import { AppConfig, AppFramework, AppType, app } from '@pixas/common';
import { IBundler, StartOptions } from './types';
export * from './types';

export default abstract class Bundler implements IBundler {
  public app: AppConfig;
  public type: AppType;
  public framework: AppFramework;
  constructor() {
    this.app = app.config;
    this.type = app.config.type || 'spa';
    this.framework = app.config.framework || 'react';
  }

  /**
   * Start a development server
   */
  public abstract start(options?: StartOptions): Promise<void> | void;

  /**
   * Build the project
   */
  public abstract build(): Promise<void> | void;
}
