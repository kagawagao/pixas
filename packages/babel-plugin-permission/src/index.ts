import { PluginObj } from '@babel/core';
import { declare } from '@babel/helper-plugin-utils';
import babelPluginSyntaxJsx from '@babel/plugin-syntax-jsx';
import { defaultProps, defaultRemoveProps, defaultSymbols } from './constants';
import { ExtractedPermissionDescriptor, Options, PermissionPluginPass } from './types';
import { visitor as CallExpression } from './visitors/expression';
import { visitor as JSXOpeningElement } from './visitors/jsx';

export type ExtractionResult<M = Record<string, string>> = {
  permissions: ExtractedPermissionDescriptor[];
  meta: M;
};

export default declare<Options, PluginObj<PermissionPluginPass>>((api, options) => {
  api.assertVersion(7);
  const {
    onPermissionExtracted,
    symbols = defaultSymbols,
    properties = defaultProps,
    removeProperties = defaultRemoveProps,
  } = options;
  const componentNames = new Set<string>([...(symbols.components ?? []), ...defaultSymbols.components]);
  const singleFunctionNames = new Set<string>([...(symbols.single ?? []), ...defaultSymbols.single]);
  const multipleFunctionNames = new Set<string>([...(symbols.multiple ?? []), ...defaultSymbols.multiple]);
  const props = new Set<string>([...(properties ?? []), ...defaultProps]);
  const removeProps = new Set<string>([...(removeProperties ?? []), ...defaultRemoveProps]);

  return {
    inherits: babelPluginSyntaxJsx,
    pre() {
      this.componentNames = Array.from(componentNames);
      this.singleFunctionNames = Array.from(singleFunctionNames);
      this.multipleFunctionNames = Array.from(multipleFunctionNames);
      this.props = Array.from(props);
      this.removeProps = Array.from(removeProps);
      this.permissions = [];
    },

    visitor: {
      Program: {
        exit(
          _,
          {
            file: {
              opts: { filename },
            },
          },
        ) {
          if (typeof onPermissionExtracted === 'function') {
            onPermissionExtracted(filename || '', this.permissions as any);
          }
        },
      },
      JSXOpeningElement,
      CallExpression,
    },
  };
});
