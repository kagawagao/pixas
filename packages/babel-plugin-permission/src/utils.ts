import { NodePath } from '@babel/core';
import * as t from '@babel/types';
import { extractedSymbol } from './constants';
import { ExtractedPermissionDescriptor, PermissionDescriptor, PermissionDescriptorPath } from './types';

function evaluatePath(path: NodePath<any>): string {
  const evaluated = path.evaluate();
  if (evaluated.confident) {
    return evaluated.value;
  }

  throw path.buildCodeFrameError('[User Permission] Permissions must be statically evaluate-able for extraction.');
}

export function getPermissionDescriptorKey(path: NodePath<any>) {
  if (path.isIdentifier() || path.isJSXIdentifier()) {
    return path.node.name;
  }

  return evaluatePath(path);
}

function getPermissionDescriptorValue(
  path?: NodePath<t.StringLiteral> | NodePath<t.JSXExpressionContainer> | NodePath<t.TemplateLiteral>,
) {
  if (!path) {
    return '';
  }
  if (path.isJSXExpressionContainer()) {
    path = path.get('expression') as NodePath<t.StringLiteral>;
  }

  // Always trim the Permission Descriptor values.
  const descriptorValue = evaluatePath(path);

  return descriptorValue;
}

export function createPermissionDescriptor(
  propPaths: [
    NodePath<t.JSXIdentifier> | NodePath<t.Identifier>,
    NodePath<t.StringLiteral> | NodePath<t.JSXExpressionContainer>,
  ][],
  properties: string[],
): PermissionDescriptorPath {
  const defaultDescriptor = Object.keys(properties).reduce((prev, cur) => {
    prev[cur] = undefined;
    return prev;
  }, {});
  return propPaths.reduce((hash: PermissionDescriptorPath, [keyPath, valuePath]) => {
    const key = getPermissionDescriptorKey(keyPath) as keyof PermissionDescriptorPath;

    if (properties.includes(key)) {
      hash[key as keyof PermissionDescriptor] = valuePath as NodePath<t.StringLiteral>;
    }

    return hash;
  }, defaultDescriptor);
}

export function evaluatePermissionDescriptor(descriptorPath: PermissionDescriptorPath, properties: string[]) {
  const descriptor: Partial<PermissionDescriptor> = {};

  properties.forEach((key) => {
    const value = getPermissionDescriptorValue(descriptorPath[key]);

    descriptor[key] = typeof value === 'string' ? value.trim() : value;
  });

  return descriptor;
}

export function tagAsExtracted(path: NodePath<any>) {
  path.node[extractedSymbol] = true;
}

export function wasExtracted(path: NodePath<any>) {
  return !!path.node[extractedSymbol];
}

export function storePermission(
  permissionDescriptor: PermissionDescriptor,
  permissions: ExtractedPermissionDescriptor[],
) {
  permissions.push(permissionDescriptor);
}
