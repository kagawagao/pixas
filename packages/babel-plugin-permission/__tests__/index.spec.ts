import { transformFileSync } from '@babel/core';
import fs from 'node:fs';
import path from 'node:path';
import plugin from '../src';
import { ExtractedPermissionDescriptor, Options } from '../src/types';

let cacheBust = 1;

function transform(filePath: string, options: Options = {}, { multiplePasses = false } = {}) {
  function getPluginConfig() {
    return [plugin, options, Date.now() + '' + ++cacheBust];
  }

  return transformFileSync(filePath, {
    presets: [
      [
        '@babel/preset-env',
        {
          targets: {
            node: '14',
            esmodules: true,
          },
          modules: false,
          useBuiltIns: false,
          ignoreBrowserslistConfig: true,
        },
      ],
      '@babel/preset-react',
      '@babel/preset-typescript',
    ],
    plugins: multiplePasses ? [getPluginConfig(), getPluginConfig()] : [getPluginConfig()],
  });
}

function transformAndCheck(fn: string, opts: Options = {}) {
  const tsxFilePath = path.join(__dirname, 'fixtures', `${fn}/index.tsx`);
  const tsFilePath = path.join(__dirname, 'fixtures', `${fn}/index.ts`);
  const filePath = fs.existsSync(tsxFilePath) ? tsxFilePath : tsFilePath;
  const permissions: ExtractedPermissionDescriptor[] = [];

  const transformed = transform(filePath, {
    ...opts,
    onPermissionExtracted(_, items) {
      permissions.push(...items);
    },
  });
  if (transformed) {
    const { code } = transformed;
    expect({
      data: { permissions },
      code: code?.trim(),
    });
  }
}

describe('babel-plugin-user-permission', () => {
  test('UserPermissionAuthorize', function () {
    transformAndCheck('UserPermissionAuthorize');
  });

  test('definePermission', function () {
    transformAndCheck('definePermission');
  });

  test('definePermissions', function () {
    transformAndCheck('definePermissions');
  });

  test('CustomUserPermissionAuthorize', function () {
    transformAndCheck('CustomUserPermissionAuthorize', {
      symbols: {
        components: ['CustomUserPermissionAuthorize'],
      },
    });
  });

  test('customDefinePermission', function () {
    transformAndCheck('customDefinePermission', {
      symbols: {
        single: ['customDefinePermission'],
      },
    });
  });

  test('customDefinePermissions', function () {
    transformAndCheck('customDefinePermissions', {
      symbols: {
        multiple: ['customDefinePermissions'],
      },
    });
  });
});
