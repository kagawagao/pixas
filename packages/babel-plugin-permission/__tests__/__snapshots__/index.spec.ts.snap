// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin-user-permission CustomUserPermissionAuthorize 1`] = `
{
  "code": ""use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _userPermission = require("@pixas/user-permission");
(0, _userPermission.definePermission)({
  code: 'user-permission-test-define'
});
var _default = () => /*#__PURE__*/React.createElement(_userPermission.CustomUserPermissionAuthorize, {
  code: "user-permission-test"
}, /*#__PURE__*/React.createElement("button", null, "test"));
exports.default = _default;",
  "data": {
    "permissions": [
      {
        "code": "user-permission-test-define",
        "data": "",
        "description": "用户权限",
        "groups": [
          "d",
        ],
        "name": "用户权限定义测试",
        "parent": "",
        "type": "",
        "uris": "",
        "value": "",
      },
      {
        "code": "user-permission-test",
        "data": "",
        "description": "用户权限",
        "groups": [
          "d",
        ],
        "name": "用户权限测试",
        "parent": "",
        "type": "",
        "uris": "",
        "value": "",
      },
    ],
  },
}
`;

exports[`babel-plugin-user-permission UserPermissionAuthorize 1`] = `
{
  "code": ""use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _userPermission = require("@pixas/user-permission");
(0, _userPermission.definePermission)({
  code: 'user-permission-test-define'
});
var _default = () => /*#__PURE__*/React.createElement(_userPermission.UserPermissionAuthorize, {
  code: "user-permission-test"
}, /*#__PURE__*/React.createElement("button", null, "test"));
exports.default = _default;",
  "data": {
    "permissions": [
      {
        "code": "user-permission-test-define",
        "data": "",
        "description": "用户权限",
        "groups": [
          "d",
        ],
        "name": "用户权限定义测试",
        "parent": "",
        "type": "",
        "uris": "",
        "value": "",
      },
      {
        "code": "user-permission-test",
        "data": "",
        "description": "用户权限",
        "groups": [
          "d",
        ],
        "name": "用户权限测试",
        "parent": "",
        "type": "",
        "uris": "",
        "value": "",
      },
    ],
  },
}
`;

exports[`babel-plugin-user-permission customDefinePermission 1`] = `
{
  "code": ""use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UsedWithDefinedPermission = void 0;
var _userPermission = require("@pixas/user-permission");
var UsedWithDefinedPermission = () => {
  var permission = (0, _userPermission.customDefinePermission)({
    code: 'user-permission-test-define'
  });
  return /*#__PURE__*/React.createElement(_userPermission.UserPermissionAuthorize, permission, /*#__PURE__*/React.createElement("button", null, "test"));
};
exports.UsedWithDefinedPermission = UsedWithDefinedPermission;",
  "data": {
    "permissions": [
      {
        "code": "user-permission-test-define",
        "data": "",
        "description": "用户权限",
        "groups": [
          "a",
          "b",
        ],
        "name": "用户权限定义测试",
        "parent": "",
        "type": "MENU",
        "uris": "",
        "value": "",
      },
    ],
  },
}
`;

exports[`babel-plugin-user-permission customDefinePermissions 1`] = `
{
  "code": ""use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UsedWithDefinedPermissions = void 0;
var _userPermission = require("@pixas/user-permission");
var UsedWithDefinedPermissions = () => {
  var permissions = (0, _userPermission.customDefinePermissions)({
    btn: {
      code: 'user-permission-test-define-permissions'
    }
  });
  return /*#__PURE__*/React.createElement(_userPermission.UserPermissionAuthorize, permissions.btn, /*#__PURE__*/React.createElement("button", null, "test"));
};
exports.UsedWithDefinedPermissions = UsedWithDefinedPermissions;",
  "data": {
    "permissions": [
      {
        "code": "user-permission-test-define-permissions",
        "data": "",
        "description": "用户权限",
        "groups": [
          "c",
        ],
        "name": "用户权限定义测试",
        "parent": "",
        "type": "",
        "uris": "",
        "value": "",
      },
    ],
  },
}
`;

exports[`babel-plugin-user-permission definePermission 1`] = `
{
  "code": ""use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UsedWithDefinedPermission = void 0;
var _userPermission = require("@pixas/user-permission");
var UsedWithDefinedPermission = () => {
  var permission = (0, _userPermission.definePermission)({
    code: 'user-permission-test-define'
  });
  return /*#__PURE__*/React.createElement(_userPermission.UserPermissionAuthorize, permission, /*#__PURE__*/React.createElement("button", null, "test"));
};
exports.UsedWithDefinedPermission = UsedWithDefinedPermission;",
  "data": {
    "permissions": [
      {
        "code": "user-permission-test-define",
        "data": "",
        "description": "用户权限",
        "groups": [
          "a",
          "b",
        ],
        "name": "用户权限定义测试",
        "parent": "",
        "type": "MENU",
        "uris": "",
        "value": "",
      },
    ],
  },
}
`;

exports[`babel-plugin-user-permission definePermissions 1`] = `
{
  "code": ""use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UsedWithDefinedPermissions = void 0;
var _userPermission = require("@pixas/user-permission");
var UsedWithDefinedPermissions = () => {
  var permissions = (0, _userPermission.definePermissions)({
    btn: {
      code: 'user-permission-test-define-permissions'
    }
  });
  return /*#__PURE__*/React.createElement(_userPermission.UserPermissionAuthorize, permissions.btn, /*#__PURE__*/React.createElement("button", null, "test"));
};
exports.UsedWithDefinedPermissions = UsedWithDefinedPermissions;",
  "data": {
    "permissions": [
      {
        "code": "user-permission-test-define-permissions",
        "data": "",
        "description": "用户权限",
        "groups": [
          "c",
        ],
        "name": "用户权限定义测试",
        "parent": "",
        "type": "",
        "uris": "",
        "value": "",
      },
    ],
  },
}
`;

exports[`babel-plugin-user-permission selfDefineUtils 1`] = `
{
  "code": ""use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.definePermissions = exports.definePermission = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * 权限定义帮助函数
 * @param permission 权限信息
 */
var definePermission = permission => {
  var createdPermission = _objectSpread({}, permission);
  if (!createdPermission.type) {
    createdPermission.type = 'OPERATE';
  }
  return createdPermission;
};

/**
 * 权限批量定义帮助函数
 * @param permissions 权限集
 */
exports.definePermission = definePermission;
var definePermissions = definePermissions => {
  var temp = {};
  Object.keys(definePermissions).forEach(key => {
    temp[key] = definePermission(definePermissions[key]);
  });
  return temp;
};
exports.definePermissions = definePermissions;",
  "data": {
    "permissions": [],
  },
}
`;
